src/localization/online_localizer/online_localizer.cpp

```diff
@@ -32,91 +37,86 @@
-  if (!vis) {
-    printf("[ERROR][OnlineLocalizer] Visualizer is not set, but wanted!\n");
-    return false;
+  for (const auto &match : matches) {
+    image_sequence_localizer::MatchingResult::Match *match_proto =
+        matching_result_proto.add_matches();
+    match_proto->set_query_id(match.quId);
+    match_proto->set_ref_id(match.refId);
+    match_proto->set_real(match.state == NodeState::HIDDEN ? 0 : 1);
   }
-  _vis = vis;
-  return true;
-}
 
-bool OnlineLocalizer::setExpansionRate(double rate) {
-  if (rate < 0.0) {
-    printf("[ERROR][OnlineLocalizer] Wrong value for the expanstion rate. The "
-           "value should be in [0,1].\n");
-    return false;
+  std::fstream out(protoFilename,
+                   std::ios::out | std::ios::trunc | std::ios::binary);
+  if (!matching_result_proto.SerializeToOstream(&out)) {
+    LOG(ERROR) << "Couldn't open the file " << protoFilename;
+    LOG(ERROR) << "The path is NOT saved.";
+    return;
   }
-  _expansionRate = rate;
-  return true;
+  out.close();
+  LOG(INFO) << "The path was written to " << protoFilename;
 }
 
-bool OnlineLocalizer::setNonMatchingCost(double non_match) {
-  // if (non_match < 1.0) {
-  if (non_match < 0.0) {
-    printf("[ERROR][OnlineLocalizer] Invalid Matching cost\n");
-    return false;
-  }
-  _nonMatchCost = non_match;
-  return true;
+OnlineLocalizer::OnlineLocalizer(SuccessorManager *successorManager,
+                                 double expansionRate, double nonMatchingCost) {
+
+  CHECK(successorManager) << "Successor manager is not set.";
+  CHECK(expansionRate > 0 && expansionRate <= 1)
+      << "Graph expansion rate should be in [0,1]. Obtained: " << expansionRate;
+  CHECK(nonMatchingCost > 0)
+      << "Non matching cost should be > 0. Obtained: " << nonMatchingCost;
+
+  successorManager_ = successorManager;
+  expansionRate_ = expansionRate;
+  nonMatchingCost_ = nonMatchingCost;
+
+  pred_[SOURCE_NODE.quId][SOURCE_NODE.refId] = SOURCE_NODE;
+  Node source = SOURCE_NODE;
+  source.accCost = 0.0;
+  accCosts_[SOURCE_NODE.quId][SOURCE_NODE.refId] = source.accCost;
+  frontier_.push(source);
+  currentBestHyp_ = source;
 }
 
-bool OnlineLocalizer::isReady() const {
-  if (!successorManager_) {
-    printf("[ERROR][OnlineLocalizer] Successor manager is not set\n");
-    return false;
-  }
-  if (_querySize == 0) {
-    printf("[ERROR][OnlineLocalizer] Size of the query sequence is not set\n");
-    return false;
-  }
-  if (_expansionRate < 0.0) {
-    printf("[ERROR][OnlineLocalizer] Expansion rate is not set\n");
-    return false;
+Matches OnlineLocalizer::findMatchesTill(int queryId) {
+  CHECK(queryId >= 0) << "Number of queries is <= 0: " << queryId;
+  Timer timer;
+  // For the first image consider lost
+  // for every image in the query set
+  for (int qu = 0; qu < queryId; ++qu) {
+    // while the graph is not expanded till row 'qu'
+    timer.start();
+    processImage(qu);
```

Comment:
If this does what the above comment says, maybe a better name for this function would be `processImageUntil` or smth along these lines? Then you wouldn't need a comment.